### DISCUSSION TITLE
Very slight bug report

### DISCUSSION THREAD

**Discussion by vkichline** [thehighnibble/thehighnibble.github.io#143] [IMSAI8080esp replica]
I built my kit last week; excellent kit and very high quality software. I really feel like I got my money's worth.
I have found one very slight issue. Perhaps this even repro's on real hardware, I don't know.
Using configuration setting $0958 (CP/M, MPU-A ROM + VIO ROM + Wifi Station) I transfer the console to TTY using: PIP CON:=TTY:.
If I then press Ctrl-P to direct output to the printer, I just get two of every character typed and nothing goes to the printer.
If I direct the console back to CRT:, printing works normally and characters are not doubled.
I don't require a fix for this, there is an easy work-around, but I thought it would be worth mentioning. If this is an issue that repro'd with the original hardware it would be really interesting to know why.

**Comment by donfroula** [comment]
Although the first serial port seems locked to the TTY console, it is possible to set the second serial port to device "UC1" at 9600 baud with the following lines in boot.conf:

UART1=9600,cs8
...

SIO2.portA.device=UART1
SIO2.portB.device=UART1
(Note if BOTH SIO2.portA.device and SIO2.portB.device are not set to UART1, the machine will crash when using device UC1 in CP/M.)

Be sure the jumpers are set correctly for the serial port connections. 

I have an Epson TMU325D serial 40 column receipt printer connected to the second serial port. I can send a text file to the printer with:

PIP UC1:=FILE.TXT

Unfortunately, the BIOS in the emulation does not support setting the LST: printer device to UC1 in the IOBYTE mapping. The following does NOT work:

STAT LST:=UC1:

**Comment by udo-munk** [comment]
To show the current I/O assignments that CP/M uses:

A>stat dev:
CON: is TTY:
RDR: is PTR:
PUN: is PTP:
LST: is LPT:

To show the possible assignments:

A>stat val:

Temp R/O Disk: d:=R/O
Set Indicator: d:filename.typ $R/O $R/W $SYS $DIR
Disk Status  : DSK: d:DSK:
User Status  : USR:
Iobyte Assign:
CON: = TTY: CRT: BAT: UC1:
RDR: = TTY: PTR: UR1: UR2:
PUN: = TTY: PTP: UP1: UP2:
LST: = TTY: CRT: LPT: UL1:

So, to assign the printer to the second serial do:
stat lst:=ul1:

And of course use stat to assign the console to tty or crt.

**Reply by donfroula** [reply]
The problem is, the CP/M version supplied with the replica does not have UL1 mapped to the second serial port, so "stat lst:=ul1:" will not work. UC1 IS mapped to the second port, but STAT LST:=UC1" is invalid, as it is not in the list of possible assignments.

However, I did find a workaround. By connecting the serial printer to the first serial port (mapped to TTY) and setting the correct baud rate in the boot configuration, I'm able to do a "STAT LST:=TTY" successfully. The downside is that boot console output directs to the printer on this port, but just leaving the printer switched off during a boot of the ESP solves this. Also, I must use the video console and not the Web/S132 TTY for the console. Now I can CTL-P to toggle printing of console output on and off. The second port is free for use in BASIC for connection to my speech synthesizer.

Best,

Don

**Comment by udo-munk** [comment]
In CP/M 2.2 version B03 ul1: is assigned to the second serial port, in versions before that might not be the case, don't know anymore. Disk image with B03 is available here: 
https://github.com/udo-munk/z80pack/tree/master/imsaisim/disks/library

**Reply by donfroula** [reply]
I downloaded the B03 version from the link provided, but it would not boot on the replica. I'm guessing there are some BIOS differences in that version, as it is likely compiled for the computer-based sim. It does work on my Raspberry Pi version of the IMSAI machine.

I guess I am stuck with my workaround for now, unless I want to go down that rabbit hole!

Thanks for your replies.

**Comment by udo-munk** [comment]
Hm, the OS's should work identical on the ESP replica and the workstations. I always use a workstation of course to build some disk images and copy them to the ESP MicroSD. Just checked on a PC, it is a 54 KB CP/M 2.2 that works ok with TTY and CRT and all boot ROM's. No idea why it won't work on your replica, just copied to mine and works.

**Reply by donfroula** [reply]
I tried updating the library on the ESP replica once again with the updated image. I mounted it in Drive A and tried a reset. I was getting some weird results, including random strings and lack of response to certain commands. I tried a full machine reset and had better results, but still some weird output like a continuous CR/LF stream. Eventually it settled down and became reliable.

I did verify that UL1 is, indeed, mapped to the second serial port rather than UC1 with the supplied 2.2 disk image.

I can assign LST to UL1 and have full printer support on the second serial port.

**Reply by udo-munk** [reply]
The B03 BIOS is a bit larger than previous version, more and better device support. So just changing disk with a different OS than what was booted won't work, jumps of the rails with unpredictable results.

Such continuous streams come when the UART buffer in the ESP32 was overrun, then it goes berserk. A complete power cycle will clean up things again.

Excellent, glad you got it working.

**Reply by donfroula** [reply]
![IMG_1621](https://github.com/user-attachments/assets/16510097-db29-47cd-a4fe-b2f7e109096b)

![IMG_1622](https://github.com/user-attachments/assets/798fe2d6-da8a-4c3e-820d-109ff35a9457)

https://github.com/user-attachments/assets/061ac9c4-bed5-4a74-85b4-156cf631f81e

**Reply by udo-munk** [reply]
Nice, ahhh that saw like noises from the matrix printer ;-)

**Comment by udo-munk** [comment]
And also with CP/M 3 you can assign TTY2 to LST with the device command. Syntax is LST:=TTY2

**Comment by donfroula** [comment]
The nice thing about CP/M 2.2 vB03 is that with UL1 now mapped to the second serial port UART1 (at SIO2.portA, if so configured in boot.conf) for connection to a serial printer, UC1 is available to Kermit to be used to connect to the virtual Telnet Modem (at SIO2.portB, if so configured in boot.conf).

In the CP/M 2.2 vB01 version, UC1 seems to be mapped in BIOS to both the Modem and UART1, preventing use of the Modem in Kermit if UC1 is otherwise assigned to the console with stat "CON:=UC1".

I'm having lots of fun with the virtual modem that takes up no physical serial ports, yet allowing connection to hundreds of BBS systems via Telnet. Use of the modem should be documented a bit better, though (only upper case AT commands allowed, etc.)!

Don

**Comment by udo-munk** [comment]
I needed a while to implement BIOS functions for all physical devices and the mappings to the logical devices, so that such things are possible. You see that at the build number of the BIOS and you really should update to B03 if not done already.

Well, the modem is a pretty basic AT modem, nothing fancy, a manual of one of the early AT modems from Hayes or so probably will do. If I remember right AT commands always were upper case only.

**Reply by donfroula** [reply]
How soon we forget. I had one of the earliest Hayes 1200 modems I used it with a Commodore 64 running a VT100 emulator to dial into the Motorola mainframe back in the early 1980s when I first started working there. I still keep a few around for remote programming of payphones and other odd applications. I built a current loop interface adapter and had an ASR33 Teletype connected to the C64, as the C64's software UART supported 110 baud and two stop bits. Worked great.

I noticed that handshaking is not configurable on the physical ports in boot.conf. With CP/M 2.2b03, does the BIOS play any role in how UL1 responds to hardware handshaking or XON/XOFF codes? If I configure LST to UL1, and turn on LST echoing of the console with CTL-P, I can overrun the buffer with a large screen dump to the serial port. The printer does send XOFF when its buffer fills, but CP/M ignores it. 

I think one of the ports associated with the serial ports (Port 37 for SIO 2 Channel B) shows the status of the hardware handshaking lines, but have not seen any documentation for the IMSAI8080esp on how the bits are mapped.

I'm thinking Kermit would handle XON/XOFF better than CP/M.

**Comment by udo-munk** [comment]
No, the BIOS doesn't implement any flow control or handshaking. I have no hardware for testing this, needs to be implemented by someone who has everything needed to get this working. Also I don't know if the ESP firmware does anything with the handshaking lines.
